shiny/ui/_theme.py

```diff
@@ -0,0 +1,513 @@
+        **kwargs
+            Keyword arguments containing Sass value declarations to add. The keys
+            should be Sass variable names using underscore casing that will be
+            transformed automatically to kebab-case. For example,
+            `.add_mixins(primary_color="#ff0000")` is equivalent to
+            `.add_mixins("$primary-color: #ff0000;")`.
+        """
+        mixins = self._combine_args_kwargs(*args, kwargs=kwargs)
+        self._mixins.extend(mixins)
+        self._reset_css()
+        return self
+
+    def add_rules(
+        self: T,
+        *args: str,
+        **kwargs: str | float | int | bool | None,
+    ) -> T:
+        """
+        Add custom Sass rules to the theme.
+
+        Sass code added via this method will be placed **after** the rule declarations
+        from the theme preset, allowing you to override or extend the default rules.
+
+        Parameters
+        ----------
+        *args
+            Sass code, as a single or multiple strings, containing rules to add.
+        **kwargs
+            Keyword arguments containing Sass value declarations to add. The keys
+            should be Sass variable names using underscore casing that will be
+            transformed automatically to kebab-case. For example,
+            `.add_rules(primary_color="#ff0000")` is equivalent to
+            `.add_rules("$primary-color: #ff0000;")`.
+        """
+        rules = self._combine_args_kwargs(*args, kwargs=kwargs)
+        self._rules.extend(rules)
+        self._reset_css()
+        return self
+
+    def to_sass(self) -> str:
+        """
+        Returns the custom theme as a single Sass string.
+
+        Returns
+        -------
+        :
+            The custom theme as a single Sass string.
+        """
+        path_functions = path_pkg_preset(self._preset, "_01_functions.scss")
+        path_defaults = path_pkg_preset(self._preset, "_02_defaults.scss")
+        path_mixins = path_pkg_preset(self._preset, "_03_mixins.scss")
+        path_rules = path_pkg_preset(self._preset, "_04_rules.scss")
+
+        sass_lines = [
+            f'@import "{path_functions}";',
+            *self._functions,
+            *self._defaults,
+            f'@import "{path_defaults}";',
+            f'@import "{path_mixins}";',
+            *self._mixins,
+            f'@import "{path_rules}";',
+            *self._rules,
+        ]
+
+        return "\n".join(sass_lines)
+
+    def to_css(
+        self,
+        compile_args: Optional[SassCompileArgs] = None,
+    ) -> str:
+        """
+        Compile the theme to CSS and return the result as a string.
+
+        Returns
+        -------
+        :
+            The compiled CSS for the theme. The value is cached such that previously
+            compiled themes are returned immediately. Adding additional custom Sass code
+            or changing the preset will invalidate the cache.
+        """
+        if self._css:
+            return copy.copy(self._css)
```

Comment:
Why is copying needed (strings are immutable)?

Reply:
This fixed an error when I added tests, but I think it was really masking another issue that I've since fixed. You're right that it's not needed and I'll take it out.